apps/labs/posts/unlocking-c-level-performance-in-df-apply.md

```diff
@@ -0,0 +1,463 @@
+(10,000 rows and 3 columns), and we will have a mix of randomly generated integers and floats.
+
+We will perform 4 tests here on:
+
+1. Analyzing overhead of the numba engine
+2. Indexing performance of the numba engine
+3. Analyzing performance of the numba engine on normalizing rows of a DataFrame
+4. Comparing raw=True and raw=False, and analyzing the speedups provided by parallel execution with the numba engine when operating on the raw values of the DataFrame.
+
+<details>
+    <summary>DataFrame Creation Code</summary>
+
+```python
+import pandas as pd
+import numpy as np
+
+np.random.seed(42)
+
+size = int(10**4)
+
+df = pd.DataFrame({"a": np.random.randn(size), "b": np.random.randn(size), "c": np.random.randn(size)})
+```
+
+</details>
+
+```python
+df
+```
+
+<div>
+  <table border="1" class="dataframe">
+    <thead>
+      <tr>
+        <th></th>
+        <th>a</th>
+        <th>b</th>
+        <th>c</th>
+      </tr>
+    </thead>
+    <tbody>
+      <tr>
+        <th>0</th>
+        <td>0.496714</td>
+        <td>-0.678495</td>
+        <td>0.348286</td>
+      </tr>
+      <tr>
+        <th>1</th>
+        <td>-0.138264</td>
+        <td>-0.305499</td>
+        <td>0.283324</td>
+      </tr>
+      <tr>
+        <th>2</th>
+        <td>0.647689</td>
+        <td>-0.597381</td>
+        <td>-0.936520</td>
+      </tr>
+      <tr>
+        <th>3</th>
+        <td>1.523030</td>
+        <td>0.110418</td>
+        <td>0.579584</td>
+      </tr>
+      <tr>
+        <th>4</th>
+        <td>-0.234153</td>
+        <td>1.197179</td>
+        <td>-1.490083</td>
+      </tr>
+      <tr>
+        <th>...</th>
+        <td>...</td>
+        <td>...</td>
+        <td>...</td>
+      </tr>
+      <tr>
+        <th>9995</th>
+        <td>1.301102</td>
+        <td>0.387827</td>
+        <td>-0.577288</td>
+      </tr>
+      <tr>
+        <th>9996</th>
+        <td>-1.998345</td>
+        <td>1.377367</td>
+        <td>1.444604</td>
+      </tr>
+      <tr>
+        <th>9997</th>
+        <td>-0.705317</td>
+        <td>0.378354</td>
+        <td>-0.501367</td>
+      </tr>
+      <tr>
+        <th>9998</th>
+        <td>0.495766</td>
+        <td>1.713530</td>
+        <td>-1.892357</td>
+      </tr>
+      <tr>
+        <th>9999</th>
+        <td>0.644388</td>
+        <td>-1.619920</td>
+        <td>-0.228884</td>
+      </tr>
+    </tbody>
+  </table>
+  <p>10000 rows × 3 columns</p>
+</div>
+
+#### Measuring compilation and boxing/unboxing overhead
+
+First, to give an idea of how much overhead compilation and boing/unboxing creates,
+we will use a simple function that returns the input Series without modification.
+
+```python
+f = lambda x: x
+```
+
+```python
+# Timing the compilation time
+# To estimate this, we will use the numba engine on a very small DataFrames
+small_df = pd.DataFrame({"a": [1]})
+%time small_df.apply(f, engine="numba", axis=1)
+```
+
+    CPU times: user 312 ms, sys: 6.93 ms, total: 319 ms
```

Comment:
This doesn't seem to render properly. (It just renders as plain text)

Is there a markdown tag that renders this not as text or code but as console output?

Reply:
I updated it to code blocks. I'm thinking about using IPython blocks for the rendering (since you can't use magics like timeit in a regular Python repl)

Does that still render fine?